pkg/types/builder.go

```diff
@@ -127,6 +131,11 @@ func (g *Builder) AddToBuilder(typeNames *TypeNames, r *resource) (*types.Named,
 	obsType := types.NewNamed(typeNames.ObservationTypeName, types.NewStruct(r.obsFields, r.obsTags), nil)
 	g.genTypes = append(g.genTypes, obsType)
 
+	for _, p := range r.topLevelRequiredParams {
+		g.validationRules += "\n"
+		g.validationRules += fmt.Sprintf(`// +kubebuilder:validation:XValidation:rule="self.managementPolicy == 'ObserveOnly' || has(self.forProvider.%s)",message="%s is a required parameter"`, p, p)
```

Comment:
We are switching to CEL rules instead of required tags. In marketplace documentation we have required marks for fields. In the new case, we need to handle this situation of the field documentation.

Reply:
Great catch!

Opened an issue to track that: https://github.com/upbound/shimmer/issues/531